x-pack/plugins/observability_ai_assistant/public/functions/recall.ts

```diff
@@ -72,28 +76,214 @@ export function registerRecallFunction({
+    Please give a score between 1 and 7, fractions are allowed.
+    A higher score means it is more relevant.`);
+  const extendedSystemMessage = {
+    ...systemMessage,
+    message: {
+      ...systemMessage.message,
+      content: `${systemMessage.message.content}\n\n${systemMessageExtension}`,
+    },
+  };
+
+  const userMessageOrQueries =
+    userMessage && userMessage.message.content ? userMessage.message.content : queries.join(',');
+
+  const newUserMessageContent =
+    dedent(`Given the question "${userMessageOrQueries}", can you give me a score for how relevant the following documents are?
+
+  ${JSON.stringify(suggestions, null, 2)}`);
+
+  const newUserMessage: Message = {
+    '@timestamp': new Date().toISOString(),
+    message: {
+      role: MessageRole.User,
+      content: newUserMessageContent,
+    },
+  };
+
+  const scoreFunction = {
+    name: 'score',
+    description:
+      'Use this function to score documents based on how relevant they are to the conversation.',
+    parameters: {
+      type: 'object',
+      additionalProperties: false,
+      properties: {
+        scores: {
+          description: 'The document IDs and their scores',
+          type: 'array',
+          items: {
+            type: 'object',
+            additionalProperties: false,
+            properties: {
+              id: {
+                description: 'The ID of the document',
+                type: 'string',
+              },
+              score: {
+                description: 'The score for the document',
+                type: 'number',
+              },
             },
           },
-          signal,
-        })
-        .then((response): { content: Serializable } => ({
-          content: response.entries.map((entry) =>
-            omit(entry, 'labels', 'score', 'is_correction')
-          ) as unknown as Serializable,
-        }));
-    }
+        },
+      },
+      required: ['score'],
+    } as const,
+    contexts: ['core'],
+  };
+
+  const response = (await service.callApi('POST /internal/observability_ai_assistant/chat', {
+    params: {
+      query: {
+        stream: false,
+      },
+      body: {
+        connectorId,
+        messages: [extendedSystemMessage, newUserMessage],
+        functions: [scoreFunction],
+        functionCall: 'score',
+      },
+    },
+    signal,
+  })) as CreateChatCompletionResponse;
+
+  const scoreFunctionRequest = decodeOrThrow(scoreFunctionRequestRt)(response);
+  const { scores } = decodeOrThrow(jsonRt.pipe(scoreFunctionArgumentsRt))(
+    scoreFunctionRequest.choices[0].message.function_call.arguments
   );
+
+  if (scores.length === 0) {
+    return [];
+  }
+
+  const relevantDocumentIds = scores
+    .filter((document) => document.score > 4)
+    .sort((a, b) => b.score - a.score)
+    .slice(0, 5)
+    .map((document) => document.id);
+
+  const relevantDocuments = suggestions.filter((suggestion) =>
+    relevantDocumentIds.includes(suggestion.id)
+  );
```

Comment:
Let's move this check up before the slice. Otherwise we might end up with less documents than needed due to the LLM hallucinating.

Reply:
Fixed in [440c949](https://github.com/elastic/kibana/pull/172164/commits/440c949fad17909168abf6804c3c7eb7e3dcba9e)